### ISSUE TITLE
string formatting - multiple lines

### ISSUE THREAD

**Issue by mojc** [CyberShadow/squelch#6]
Hey!
We started using duckdb as well which has its own dialect and in general the squelch works well with it as well :tada: 
There are some edge cases such as [PRINTF](https://duckdb.org/docs/sql/functions/char.html#printf-syntax).

Currently using PRINTF as:
```
PRINTF(
  'weight: %''d
  wmean: %.2g
  wpred_mean: %.2g
  sum_kpi: %''d
  sum_pred_kpi: %''d',
       CAST(total_weight AS INT64),
       SUM(weight * kpi) / total_weight,
       SUM(weight * pred_kpi) / total_weight,
       CAST(SUM(kpi) AS INT64),
       CAST(SUM(pred_kpi) AS INT64)
     )
  AS wbias_tooltip,
```
And squelch formats it to:
```
PRINTF(
       'weight: %' 'd\n  wmean: %.2g\n  wpred_mean: %.2g\n  sum_kpi: %' 'd\n  sum_pred_kpi: %' 'd',
       CAST(total_weight AS INT64),
       SUM(weight * kpi) / total_weight,
       SUM(weight * pred_kpi) / total_weight,
       CAST(SUM(kpi) AS INT64),
       CAST(SUM(pred_kpi) AS INT64)
     )
```
With concatenating lines with `\n`.

That unfortunately does not seem to work with duckdb:
`ParserException: Parser Error: syntax error at or near "'d\n wmean: %.2g\n wpred_mean: %.2g\n sum_kpi: %'" LINE 9: 'weight: %' 'd\n wmean: %.2g\n wpred_mean: %.2g\n sum_kpi: %' 'd\n sum_pred_kpi: %' 'd', CAST(total_weight AS INT64), SUM(weight * kpi) / total_weight, SUM(weight * pred_kpi) / total_weight, CAST(SUM(kpi) AS INT64), CAST(SUM(pred_kpi) AS INT64) ) AS wbias_tooltip `

Would it be possible to skip line concatenation in PRINTF functions or in general?

**Comment by CyberShadow** [comment]
Hi :wave: 

I don't think the problem is with `PRINTF` but the string literal syntax of the dialect.

Does DuckDB understand triple quotes? I.e. would this work:

```
PRINTF(
  '''weight: %'d
  wmean: %.2g
  wpred_mean: %.2g
  sum_kpi: %'d
  sum_pred_kpi: %'d''',
  CAST(total_weight AS INT64),
  SUM(weight * kpi) / total_weight,
  SUM(weight * pred_kpi) / total_weight,
  CAST(SUM(kpi) AS INT64),
  CAST(SUM(pred_kpi) AS INT64)
) AS wbias_tooltip
```

**Comment by mojc** [comment]
Thanks for the quick reply :+1: 

I tried it add it works-ish.
The query runs successfully for that instance. I do get an `'` artefact in the final plotting:
![image](https://github.com/CyberShadow/squelch/assets/38767382/383daef1-b898-4969-8745-0b1237e062d8)
Not sure what happens here.

The other thing is that this doesn't seem to work with shorter strings it seems, e.g.:
```
PRINTF(
'weight: %''d
wsmape: %.2g',
       CAST(total_weight AS INT64),
         SUM(weight * ABS(pred_kpi - kpi) * 2 / (pred_kpi + kpi))
       / total_weight
     )
  AS wsmaspe_tooltip,
```
Which formats to:
```
     PRINTF(
       "weight: %''d\nwsmape: %.2g",
       CAST(total_weight AS INT64),
         SUM(weight * ABS(pred_kpi - kpi) * 2 / (pred_kpi + kpi))
       / total_weight
     )
  AS wsmaspe_tooltip,
```

**Comment by CyberShadow** [comment]
OK, I see. So, it doesn't support triple quotes. It understands `'''` as the beginning of a string, and then a literal apostrophe.

**Comment by CyberShadow** [comment]
How does this output look:

```
PRINTF(
  $$weight: %'d
  wmean: %.2g
  wpred_mean: %.2g
  sum_kpi: %'d
  sum_pred_kpi: %'d$$,
  CAST(total_weight AS INT64),
  SUM(weight * kpi) / total_weight,
  SUM(weight * pred_kpi) / total_weight,
  CAST(SUM(kpi) AS INT64),
  CAST(SUM(pred_kpi) AS INT64)
) AS wbias_tooltip;

PRINTF(
  'weight: %''d
wsmape: %.2g',
  CAST(total_weight AS INT64),
    SUM(weight * ABS(pred_kpi - kpi) * 2 / (pred_kpi + kpi))
  / total_weight
) AS wsmaspe_tooltip;
```

**Comment by mojc** [comment]
Yes the `$$` seem to work fine :+1: 

```
PRINTF(
  $$weight: %''d
  wmean: %.2g
  wpred_mean: %.2g
  sum_kpi: %''d
  sum_pred_kpi: %''d$$,
  CAST(total_weight AS INT64),
  SUM(weight * kpi) / total_weight,
  SUM(weight * pred_kpi) / total_weight,
  CAST(SUM(kpi) AS INT64),
  CAST(SUM(pred_kpi) AS INT64)
) AS wbias_tooltip,
```

**Comment by CyberShadow** [comment]
Great! I pushed e07a69ae1572cdcc3d20b57387a08e274424ec9f to the `next` branch. One thing that isn't clear to me is how to integrate it with `pre-commit`. Do you know if it's possible to add a `--dialect=duckdb` argument on your side?

**Comment by mojc** [comment]
This is an example of a pre-commit hook with an argument:
```
hooks:
      - id: terraform_fmt
        args: [--args=-no-color]
```
So it should work like this:
```
  - repo: https://github.com/CyberShadow/squelch
    rev: v0.2.7
    hooks:
      - id: squelch
        args: [--dialect=duckdb]
        exclude: '^onboarding/query_templates/.*\.sql|toolz/streamlit/pages/queries/utils/rollup_metrics.sql'
```

We don't currently mark the queries with a dialect but as we are using duckdb more and more seems like we might need to.

**Comment by CyberShadow** [comment]
OK, I pushed it to a new `v0.3.0` tag, can you give it a try?

**Comment by mojc** [comment]
Works for the test cases with the PRINTF but had other problems:
```
SELECT
  {%- if facet %}
    {{ facet }},
  {%- endif %}
```
formatted to:
```
{%- if facet %}
    ,
  {%- endif %}
```
Basically all `{{ var }}` disappear.

**Comment by CyberShadow** [comment]
Ah, fixed in v0.3.1.

**Comment by mojc** [comment]
that worked, thank you!
